### PR TITLE
fix: resolved offerings: calculate if a container is missing

### PR DESCRIPTION
**Pull request by g-linville**
We saw cases (involving nested acorns) where resolved offerings would have containers missing from the list entirely. I think this was due to weird behavior where the `.Status.AppStatus.Containers` would not be populated on a nested acorn the first time it ran through the resolved offerings handler. To make sure that everything ultimately gets resolved, I added this new `needsCalculation` function that checks to make sure that every container defined in the app is also present in the resolved offerings.

### Checklist
- [x] The title of this PR would make a good line in Acorn's Release Note's Changelog
- [ ] The title of this PR ends with a link to the main issue being address in parentheses, like: `This is a title (#1216)`. [Here's an example](https://github.com/acorn-io/runtime/pull/1199)
- [x] All relevant issues are referenced in the PR description. *NOTE: don't use [GitHub keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword) that auto-close issues*
- [x] Commits follow [contributing guidance](https://github.com/acorn-io/runtime/blob/main/CONTRIBUTING.md#commits)
- [ ] Automated tests added to cover the changes. If tests couldn't be added, an explanation is provided in the Verification and Testing section
- [x] Changes to user-facing functionality, API, CLI, and upgrade impacts are clearly called out in PR description
- [ ] PR has at least two approvals before merging (or a reasonable exception, like it's just a docs change)

### CODE REVIEW

File: pkg/controller/appdefinition/deploy.go [hunk 1]
```diff
@@ -629,7 +629,7 @@ func containerAnnotation(container v1.Container) string {
 }
 
 func resolvedOfferingsAnnotation(appInstance *v1.AppInstance, container v1.Container) (string, error) {
-	if resolved, exists := appInstance.Status.ResolvedOfferings.Containers[container.Name]; exists {
+	if resolved, exists := appInstance.Status.ResolvedOfferings.Containers[strings.ToLower(container.Name)]; exists {
```
Comment (g-linville) [inline pkg/controller/appdefinition/deploy.go:632]: While stepping through the debugger, I found a case here where `container.Name` was capitalized for some reason, which is why I added this `strings.ToLower`.

**Review by thedadams — APPROVED** [review]

File: pkg/controller/appdefinition/deploy.go (hunk 1 above)
Comment (StrongMonkey) [reply pkg/controller/appdefinition/deploy.go:632]: Interesting, I thought container name is not allowed to have capital name

**Review by StrongMonkey — APPROVED** [review]

**Comment by g-linville** [comment]
Sorry @thedadams and @StrongMonkey. I just requested your re-review. While working on other stuff, I figured out the real problem with `container.Name`. Sometimes it just wasn't set at all for whatever reason. So I changed up the function that creates the resolved offerings annotation to just take the name of the container as a string, and I pass it all the way down to there.

File: pkg/controller/appdefinition/deploy.go [hunk 2]
```diff
@@ -628,8 +629,9 @@ func containerAnnotation(container v1.Container) string {
 	return string(json)
 }
 
-func resolvedOfferingsAnnotation(appInstance *v1.AppInstance, container v1.Container) (string, error) {
-	if resolved, exists := appInstance.Status.ResolvedOfferings.Containers[container.Name]; exists {
+func resolvedOfferingsAnnotation(appInstance *v1.AppInstance, containerName string) (string, error) {
+	logrus.Infof("resolved offerings annotation - container name: %s", containerName)
```
Comment (StrongMonkey) [inline pkg/controller/appdefinition/deploy.go:633]: Remove this

**Review by StrongMonkey — APPROVED** [review]

File: pkg/controller/appdefinition/deploy.go (hunk 2 above)
Comment (g-linville) [reply pkg/controller/appdefinition/deploy.go:633]: oh lol thanks

Comment (g-linville) [reply pkg/controller/appdefinition/deploy.go]: done

**Review by thedadams — APPROVED** [review]

**Pull request merged**